lib/spack/spack/fetch_strategy.py

```diff
@@ -40,89 +40,127 @@
-import copy
-from functools import wraps
-from six import string_types
-from six import with_metaclass
+import sys
+
+import six
 
 import llnl.util.tty as tty
-from llnl.util.filesystem import *
 import spack
 import spack.error
 import spack.util.crypto as crypto
 import spack.util.pattern as pattern
+
+from llnl.util.filesystem import *
+from spack.util.compression import decompressor_for, extension
 from spack.util.executable import *
 from spack.util.string import *
 from spack.version import Version, ver
-from spack.util.compression import decompressor_for, extension
 
-
-#: List of all fetch strategies, created by FetchStrategy metaclass.
+#: List of all known fetch strategies
 all_strategies = []
 
 
-def _needs_stage(fun):
-    """Many methods on fetch strategies require a stage to be set
-       using set_stage().  This decorator adds a check for self.stage."""
+def fetcher(cls):
+    """Registers a fetch strategy
 
-    @wraps(fun)
-    def wrapper(self, *args, **kwargs):
-        if not self.stage:
-            raise NoStageError(fun)
-        return fun(self, *args, **kwargs)
+    Args:
+        cls (FetchStrategy): fetcher class to be registered
+    """
+    all_strategies.append(cls)
+    return cls
 
-    return wrapper
 
+def has_expanded_tree(directory):
+    """Checks if the input directory has at least one other
+    child subdirectory.
 
-class FSMeta(type):
-    """This metaclass registers all fetch strategies in a list."""
-    def __init__(cls, name, bases, dict):
-        type.__init__(cls, name, bases, dict)
-        if cls.enabled:
-            all_strategies.append(cls)
+    Args:
+         directory (path): directory where to search
 
+    Returns:
+        first subdirectory found or None
+    """
+    for p in [os.path.join(directory, f) for f in os.listdir(directory)]:
+        if os.path.isdir(p):
+            return p
+    return None
+
+
+class FetchStrategy(object):
+    """Defines the interface that all fetchers need to comply with"""
 
-class FetchStrategy(with_metaclass(FSMeta, object)):
-    """Superclass of all fetch strategies."""
-    enabled = False  # Non-abstract subclasses should be enabled.
-    required_attributes = None  # Attributes required in version() args.
+    #: Attributes required in version() args
+    required_attributes = tuple()
+    #: If set it will be used to save an archive to a specific filename
+    save_filename = None
+    #: Private cache for source_dir, initialized to None in the class
+    _source_dir = None
+    #: Private cache for expected_archive_files,
+    #: initialized to None in the class
+    _expected_archive_files = None
 
-    def __init__(self):
-        # The stage is initialized late, so that fetch strategies can be
-        # constructed at package construction time.  This is where things
-        # will be fetched.
-        self.stage = None
+    @property
+    def source_dir(self):
+        """Root directory where the source code should be expanded"""
```

Comment:
@tgamblin For this I was quite undecided between `source_dir` and `working_dir`. Let me know if you have any opinion or a third name to put into the mix.